### PR TITLE
CI: add Ninja

### PR DESCRIPTION
**Pull request by Mixaill**
CMake + Ninja is a pretty popular combination, so I added tests for it.

* ci/linux: add Ninja and exclude Ninja + GCC + Modules
* ci/windows: add Ninja and exclude Ninja + MSVC + Modules
* ci/windows: fix string test

### CODE REVIEW

**Comment by vitaut** [comment]
Thanks for the PR but I don't think that we should add tests for other generators because none of the functionality is generator-specific.

**Comment by Mixaill** [comment]
@vitaut Please take a closer look, we currently have problems with the combination of GCC+Ninja and MSVC+Ninja (marked as `BUG` in `.yml` files), which will be fixed in the future.

Since we have generator-specific behavior, it's worth running tests for them.

**Comment by vitaut** [comment]
Could you elaborate what the problem is and which generator-specific behavior you are referring to?

**Comment by Mixaill** [comment]
@vitaut currently we are failing to build modules for Ninja+GCC and Ninja+MSVC, but we are ok with Make+GCC and MSBuild+MSVC:

https://github.com/Mixaill/modules/actions/runs/4877248308/jobs/8701696712
https://github.com/Mixaill/modules/actions/runs/4877248307/jobs/8701696650

So, since we have different behavior for Ninja, Make and MSBuild, it should be fine to have CI for different generators because behavior between them is somehow not the same.

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -10,28 +10,35 @@ jobs:
...
         standard: [17, 20, 23]
-        build_type: [Debug]
+        # BUG: Ninja + MSVC + Modules fails to build
+        exclude:
+          - {generator: "Ninja", standard: 20 }
+          - {generator: "Ninja", standard: 23 }
 
     steps:
     - uses: actions/checkout@v3
+    - uses: ilammy/msvc-dev-cmd@v1
```
Comment (vitaut) [inline .github/workflows/windows.yml]: What is this?

File: .github/workflows/windows.yml [hunk 2]
```diff
@@ -10,28 +10,35 @@ jobs:
...
+      run: cmake --build . --config Debug --parallel
 
     - name: Test
       env:
+        FILE_PATH: ${{fromJSON('["build/Debug/main.exe", "build/main.exe"]')[matrix.generator == 'Ninja']}}
         CHECK_STRING: ${{fromJSON('["Modules are late to the party", "Hello, modules"]')[matrix.standard > 17]}}
-      run: $(${{runner.workspace}}/build/${{matrix.build_type}}/main.exe | Out-String) -match "${env:CHECK_STRING}"
+      run: |
+        $result = (& ${{runner.workspace}}\${env:FILE_PATH} | Out-String) -match "${env:CHECK_STRING}"
+        if(!$result) { Write-Error "Bad output" }
```
Comment (vitaut) [inline .github/workflows/windows.yml]: I think we should avoid scripting in yaml. I suggest moving this logic elsewhere, if possible to the binary itself or some python or even cmake script.

File: .github/workflows/windows.yml (hunk 1 above)
Comment (Mixaill) [reply .github/workflows/windows.yml]: For Ninja + MSVC you need to have `cl.exe` in the `PATH` environment variable. 

Usually people use `vcvarsall.bat` or `VsDevCmd.cmd` or specify the full path in `CMAKE_{C,CXX}_COMPILER` option. 

But for MSVS, the location of MSVC is not constant and can be changed with github workflows OS image update.

To overcome this kind of problem, Microsoft came up with `vswhere.exe`, which searches for the MSVC compiler.

It seems that `ilammy/msvc-dev-cmd` is the easiest way to use `vswhere` for github workflows.
It:
* finds `vswhere.exe` on the system
* runs it and receives info about MSVC location
* executes `vcvarsall.bat` for the last available MSVC (or for explicit version, if you specified it), so we will have `cl.exe` in the `PATH` environment variable.

File: .github/workflows/windows.yml (hunk 2 above)
Comment (Mixaill) [reply .github/workflows/windows.yml]: >or even cmake script.

image

File: .github/workflows/windows.yml (hunk 1 above)
Comment (vitaut) [reply .github/workflows/windows.yml]: MSVC + ninja is an unusual combination and considering this complexity I suggest only testing ninja on Linux.

Comment (Mixaill) [reply .github/workflows/windows.yml]: Rolling back. Perhaps Linux+Ninja will be enough, otherwise, we can come back to this later.

However, MSVC + Ninja is not unusual these days: it is the default for CLion+MSVC and Vcpkg, so the user base is actually quite large.

File: .github/workflows/windows.yml (hunk 2 above)
Comment (Mixaill) [reply .github/workflows/windows.yml]: Rewritten, now both Linux and Windows use the same CMake script. I put it in `.github/workflows` to avoid confusing people browsing the `example` folder.

**Comment by Mixaill** [comment]
@vitaut issues adressed

**Comment by vitaut** [comment]
Thanks!
